src/greatwall/gui.py

```diff
@@ -934,8 +953,10 @@ def config_selecting_derivation_widgets_layout(self):
             for idx in range(self.greatwall.tree_arity):
                 view = ImageViewer(self)
                 selection_button = QPushButton(str(idx), self)
+                button_minimize = self.init_button_minimize(view)
```

Comment:
There are many other scenarios where a minimize-maximize button that shows "-" to minimize and "+" to maximize could be useful. This code can be reused in many places and even be refactored away in its own file, maybe "buttons/minmax.py" as an example.

This leads to less code to read and less places to change. When changing duplicate code, made from copying and paste, there are now two places to update rather than one. But when we abstract away the common pattern in a shared piece of code, only that piece of code gets to be changed when needed (the client requested a new style, a slightly different behavior).

Reply:
And how my suggestion violate what you mentioned? you can pass to the connected function any parameters that you can later use to customize user options! Is it really useful to seperate the logic of the widet in different file, do you have examples of any very known projects that do this?

Also, in contrast to what you are claiming about what your code can do, it is not usefull that much; consider the following: We already have a button to show and hide the result output, could you please give me a code snippt for using your code to implement the logic of the result button? I think you will end up by list of if-statement that will complicate the mintanace.

Last thing, the copy and paste code isn't like that! Keep in your mind that we do'nt reinvent the wheel; if some thing exist in popular project and fit your need then simply use it, do you see this a copy and paste?!

Reply:
> Also, what do you think about using 'Show' and 'Hide' labels instead of '+' and '-".
> We already have a button to show and hide the result output

There's one more option: refactor so that the labels are taken as arguments when the button is built. The one calling the procedure that builds the button can decide for either "Show","Hide" or "+","-" or something else in another language.

> I think you will end up by list of if-statement that will complicate the mintanace.

It's possible to refactor so that a procedure or lambda is taken as argument, so there won't be a list of if-statements.

> do you see this a copy and paste?!

No when it's not a literal copy and paste. Projects usually provide libraries that exports procedures and constants to be used.

**Do you want me to refactor so that there's only one place where "Hide/Show" buttons are created?**

Reply:
You didn't address my questions, please read my comment again and address my questions so we can go forward, again my questions are:
1. And how my suggestion violate what you mentioned?
2. do you have examples of any very known projects that do this?
3. could you please give me a code snippt for using your code to implement the logic of the result button?